quantum/openrgb.c

```diff
@@ -0,0 +1,547 @@
+            continue;
+        }
+
+        g_openrgb_direct_mode_colors[first_led + i].r = r;
+        g_openrgb_direct_mode_colors[first_led + i].g = g;
+        g_openrgb_direct_mode_colors[first_led + i].b = b;
+    }
+
+    raw_hid_buffer[1] = OPENRGB_SUCCESS;
+    raw_hid_buffer[2] = OPENRGB_EOM;
+}
+void openrgb_direct_mode_get_led_color(uint8_t *data) {
+    const uint8_t led = data[1];
+
+    raw_hid_buffer[0] = OPENRGB_DIRECT_MODE_GET_LED_COLOR;
+
+    if (led >= DRIVER_LED_TOTAL) {
+        raw_hid_buffer[1] = OPENRGB_FAILURE;
+        raw_hid_buffer[2] = OPENRGB_EOM;
+        return;
+    }
+
+    raw_hid_buffer[1] = g_openrgb_direct_mode_colors[led].r;
+    raw_hid_buffer[2] = g_openrgb_direct_mode_colors[led].g;
+    raw_hid_buffer[3] = g_openrgb_direct_mode_colors[led].b;
+    raw_hid_buffer[4] = OPENRGB_EOM;
+}
+
+/////////////////////////////////////////////
+// Control QMK Modes
+////////////////////////////////////////////
+void openrgb_qmk_mode_set_mode_and_speed(uint8_t *data) {
+    const uint8_t mode  = data[1];
+    const uint8_t speed = data[2];
+
+    raw_hid_buffer[0] = OPENRGB_QMK_MODE_SET_MODE_AND_SPEED;
+
+    if (mode >= RGB_MATRIX_EFFECT_MAX || speed > 255) {
+        raw_hid_buffer[1] = OPENRGB_FAILURE;
+        raw_hid_buffer[2] = OPENRGB_EOM;
+        return;
+    }
+
+    rgb_matrix_sethsv_noeeprom(0, 255, 255);
+    rgb_matrix_mode_noeeprom(mode);
+    rgb_matrix_set_speed_noeeprom(speed);
+
+    raw_hid_buffer[1] = OPENRGB_SUCCESS;
+    raw_hid_buffer[2] = OPENRGB_EOM;
+}
+void openrgb_qmk_mode_set_color_mode_and_speed(uint8_t *data) {
+    const uint8_t h     = data[1];
+    const uint8_t s     = data[2];
+    const uint8_t v     = data[3];
+    const uint8_t mode  = data[4];
+    const uint8_t speed = data[5];
+
+    raw_hid_buffer[0] = OPENRGB_QMK_MODE_SET_COLOR_MODE_AND_SPEED;
+
+    if (h > 255 || s > 255 || v > 255 || mode >= RGB_MATRIX_EFFECT_MAX || speed > 255) {
+        raw_hid_buffer[1] = OPENRGB_FAILURE;
+        raw_hid_buffer[2] = OPENRGB_EOM;
+        return;
+    }
+
+    rgb_matrix_sethsv_noeeprom(h, s, v);
+    rgb_matrix_mode_noeeprom(mode);
+    rgb_matrix_set_speed_noeeprom(speed);
+
+    raw_hid_buffer[1] = OPENRGB_SUCCESS;
+    raw_hid_buffer[2] = OPENRGB_EOM;
+}
+void openrgb_qmk_mode_get_color(void) {
+    HSV hsv_color     = rgb_matrix_get_hsv();
+    raw_hid_buffer[0] = OPENRGB_QMK_MODE_GET_COLOR;
+    raw_hid_buffer[1] = hsv_color.h;
+    raw_hid_buffer[2] = hsv_color.s;
+    raw_hid_buffer[3] = hsv_color.v;
+    raw_hid_buffer[4] = OPENRGB_EOM;
+}
+
+/////////////////////////////////////////////
+// Setup OpenRGB
+/////////////////////////////////////////////
+void openrgb_get_protocol_version(void) {
+    raw_hid_buffer[0] = OPENRGB_GET_PROTOCOL_VERSION;
+    raw_hid_buffer[1] = OPENRGB_PROTOCOL_VERSION;
+    raw_hid_buffer[2] = OPENRGB_EOM;
+}
+void openrgb_get_device_name(void) {
+    uint8_t i = 0;
+    while (PRODUCT[i] != 0 && i + 2 < RAW_EPSIZE) {
```

Comment:
This will error out if there are any spaces in the name, which is VERY common.  

Same goes for manufacturer.